### PR TITLE
Support bipolar readout

### LINKED ISSUE #50: K0 limit calculation
I think the calculation of the k0 limits here are wrong:
https://github.com/ckolbPTB/mrpro/blob/52d469d7ebecc7fe5fd8cf352da8a3a0ad17241c/src/mrpro/data/_KData.py#L102

They should be calculated based on `AcqInfo.number_of_samples` and `AcqInfo.center_sample`

### LINKED ISSUE #184: Bipolar gradient
Currently we only support a single type of readout: https://github.com/PTB-MR/mrpro/blob/9b62daeb99a0bb5bb364088b4aedba37fbfdf0c8/src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py#L69

For a multi-echo acquisition with bipolar readouts the readout direction (and probably centre of the readout) is different between odd and even echos.

### PR DESCRIPTION
**Pull request by ckolbPTB**
Correct trajectory calculation for bipolar (i.e. reversed) readout.

closes #184 
closes #50 

requires #221
Update: actually does not require #217

ToDo:
- [x] Fix encoding limits for readout in KData
- [x] Add tests for bipolar readout

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Do you have a test case that fails because of #221 ?

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 1]
```diff
@@ -78,8 +81,8 @@ class DummyTrajectory(KTrajectoryCalculator):
     Shape will fit to all data. Only used as dummy for testing.
     """
 
-    def __call__(self, header: KHeader) -> KTrajectory:
-        """Calculate dummy trajectory for given KHeader."""
-        kx = torch.zeros(1, 1, 1, header.encoding_limits.k0.length)
+    def __call__(self, header: KHeader) -> KTrajectory:  # noqa: ARG002
+        """Calculate dummy trajectory."""
+        kx = torch.zeros(1, 1, 1, 1)
```
Comment (fzimmermann89) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:81]: why is this change required?

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 2]
```diff
@@ -57,19 +58,21 @@ def _kfreq(self, kheader: KHeader) -> torch.Tensor:
...
             raise ValueError('Trajectory can only be calculated if each acquisition has the same number of samples')
-        if len(torch.unique(center_sample)) > 1:
-            raise ValueError('Trajectory can only be calculated if each acquisition has the same center sample')
+        n_k0 = int(n_samples[0])
+
+        # Data can be obtained with standard or reversed readout (e.g. bipolar readout).
+        k0 = torch.linspace(0, n_k0 - 1, n_k0, dtype=torch.float32)[None, :] - center_sample[:, None]
+        reversed_readout_idx = torch.where(kheader.acq_info.flags.flatten() & AcqFlags.ACQ_IS_REVERSE.value)[0]
+        # pytorch does not yet allow for negative indexing. torch.flip makes a copy of the data. providing a reversed
+        # index should be fastest
```
Comment (fzimmermann89) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: pretty sure flip is faster and easier to read...

Comment (fzimmermann89) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: reversed_index is similiar to   reversed_readout_idx

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 3]
```diff
@@ -57,19 +58,21 @@ def _kfreq(self, kheader: KHeader) -> torch.Tensor:
...
+        center_sample = kheader.acq_info.center_sample.flatten()
         if len(torch.unique(n_samples)) > 1:
             raise ValueError('Trajectory can only be calculated if each acquisition has the same number of samples')
-        if len(torch.unique(center_sample)) > 1:
-            raise ValueError('Trajectory can only be calculated if each acquisition has the same center sample')
+        n_k0 = int(n_samples[0])
+
+        # Data can be obtained with standard or reversed readout (e.g. bipolar readout).
+        k0 = torch.linspace(0, n_k0 - 1, n_k0, dtype=torch.float32)[None, :] - center_sample[:, None]
+        reversed_readout_idx = torch.where(kheader.acq_info.flags.flatten() & AcqFlags.ACQ_IS_REVERSE.value)[0]
```
Comment (fzimmermann89) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: consider using a  boolean mask instead of torch.where:
```
reversed_readout_mask = kheader.acq_info.flags.flatten() & AcqFlags.ACQ_IS_REVERSE.value
 k0[reversed_readout_mask] = torch.flip(k0[reversed_readout_mask ])
```
or similiar should do the same and is easier to read

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 4]
```diff
@@ -57,19 +58,21 @@ def _kfreq(self, kheader: KHeader) -> torch.Tensor:
...
+        # pytorch does not yet allow for negative indexing. torch.flip makes a copy of the data. providing a reversed
+        # index should be fastest
+        reversed_index = torch.linspace(n_k0 - 1, 0, n_k0, dtype=torch.int64)
+        k0[reversed_readout_idx, :] = k0[reversed_readout_idx[:, None], reversed_index[None, :]]
 
-        # Calculate points along readout
-        n_k0 = int(n_samples[0, 0, 0])
-        k0 = torch.linspace(0, n_k0 - 1, n_k0, dtype=torch.float32) - center_sample[0, 0, 0]
-        return k0
+        return k0.reshape(kheader.acq_info.center_sample.shape[:-1] + (n_k0,))
```
Comment (fzimmermann89) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: https://docs.astral.sh/ruff/rules/collection-literal-concatenation/ does not yet catch this, as shape and the tuple are of differnt type, but you can even consider instead doing
` k0.reshape(*kheader.acq_info.center_sample.shape[:-1], n_k0)`
as reshape takes in integers

**Comment by fzimmermann89** [comment]
consider the commit a suggestion for the k0 calculation that saves a few steps

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py (hunk 3 above)
Comment (fzimmermann89) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: I checked how the synax should acutally be, see commit.

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py (hunk 4 above)
Comment (fzimmermann89) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: see commit

**Comment by fzimmermann89** [comment]
Could you also write a sentence in #50 how you solve it?

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py (hunk 1 above)
Comment (ckolbPTB) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:81]: before we set encoding_limits.k0 based on kdata.data.shape, therefore encoding_limits.k0 always matched the data. For a random header this is not the case anymore and therefore the only valid trajectory is one of size 1.

**Comment by ckolbPTB** [comment]
> Could you also write a sentence in https://github.com/PTB-MR/mrpro/issues/50 how you solve it?

done

**Review by fzimmermann89 — APPROVED** [review]

File: src/mrpro/data/_kdata/_KData.py [hunk 5]
```diff
@@ -135,7 +135,13 @@ def from_file(
...
         # Fill k0 limits if they were set to zero / not set in the header
         if kheader.encoding_limits.k0.length == 1:
-            kheader.encoding_limits.k0 = Limits(0, kdata.shape[-1] - 1, kdata.shape[-1] // 2)
+            # The encodig limits should describe the encoded space of all readout lines. If we have two readouts with
+            # (number_of_samples, center_sample) of (100, 20) (e.g. partial Fourier in the negative k0 direction) and
+            # (100, 80) (e.g. partial Fourier in the positive k0 direction) then this should lead to encoding limits of
+            # [min=0, max=159, center=80]
+            max_center_sample = int(torch.max(kheader.acq_info.center_sample))
+            max_pos_k0_extend = int(torch.max(kheader.acq_info.number_of_samples - kheader.acq_info.center_sample))
+            kheader.encoding_limits.k0 = Limits(0, max_center_sample + max_pos_k0_extend - 1, max_center_sample)
```
Comment (schuenke) [inline src/mrpro/data/_kdata/_KData.py:144]: nice example of descriptive variable names helping to understand the code 👍

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 6]
```diff
@@ -57,18 +58,17 @@ def _kfreq(self, kheader: KHeader) -> torch.Tensor:
         ValueError
             Center sample has to be the same for each readout
```
Comment (schuenke) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py]: This should be removed

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 7]
```diff
@@ -57,18 +58,17 @@ def _kfreq(self, kheader: KHeader) -> torch.Tensor:
...
-            raise ValueError('Trajectory can only be calculated if each acquisition has the same center sample')
+        n_k0 = int(n_samples.item())
 
-        # Calculate points along readout
-        n_k0 = int(n_samples[0, 0, 0])
-        k0 = torch.linspace(0, n_k0 - 1, n_k0, dtype=torch.float32) - center_sample[0, 0, 0]
+        # Data can be obtained with standard or reversed readout (e.g. bipolar readout).
+        k0 = torch.linspace(0, n_k0 - 1, n_k0, dtype=torch.float32) - center_sample
+        # Data can be obtained with standard or reversed readout (e.g. bipolar readout).
+        reversed_readout_mask = (kheader.acq_info.flags[..., 0] & AcqFlags.ACQ_IS_REVERSE.value).bool()
```
Comment (schuenke) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:68]: Just a question unrelated to the code: can `ACQ_IS_REVERSE ` be set using pulseq labels? Or would this only work with an additional meta ismrmrd file with pulseq sequences?

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py [hunk 8]
```diff
@@ -78,8 +78,8 @@ class DummyTrajectory(KTrajectoryCalculator):
     Shape will fit to all data. Only used as dummy for testing.
     """
 
-    def __call__(self, header: KHeader) -> KTrajectory:
-        """Calculate dummy trajectory for given KHeader."""
-        kx = torch.zeros(1, 1, 1, header.encoding_limits.k0.length)
+    def __call__(self, header: KHeader) -> KTrajectory:  # noqa: ARG002
```
Comment (schuenke) [inline src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:79]: why do we keep the header argument if we don't need it anymore?

**Review by schuenke — APPROVED** [review]

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py (hunk 7 above)
Comment (ckolbPTB) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:68]: Not sure what pulseq can set but this is based on a standard Siemens raw data flag, similiar to calibration data and so on.

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py (hunk 8 above)
Comment (ckolbPTB) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:79]: because otherwise the signature of the function does not match anymore

File: src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py (hunk 7 above)
Comment (fzimmermann89) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:68]: Do we need this if we use a pulseq trajectory calculator?

Otherwise, I think it will not be set...

Comment (fzimmermann89) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:68]: @schuenke
In the pulseq reference manual, label extension section, the binary flag REV is mentioned

Can you try if setting that flag in pulseq to 1 sets the reversed acquisition flag in the ismrmrd?

Comment (schuenke) [reply src/mrpro/data/traj_calculators/_KTrajectoryCalculator.py:68]: Can try it later today, yeah
